File: src/main/java/com/mkonst/analysis/MethodCallGraph.java
package com.mkonst.analysis;

import com.mkonst.types.ClassMethod;
import org.jgrapht.Graph;
import org.jgrapht.graph.DefaultDirectedGraph;
import org.jgrapht.graph.DefaultEdge;

import java.util.ArrayList;
import java.util.List;

public class MethodCallGraph {
    private Graph<ClassMethod, DefaultEdge> graph;

    public MethodCallGraph() {

        // Create a directed graph
        graph = new DefaultDirectedGraph<>(DefaultEdge.class);
    }

    /**
     * Converts the given Strings into ClassMethod instances and appends a new edge. If the graph does not contain
     * the given methods are Vertices, new vertices will be generated
     *
     * @param callerClassMethod
     * @param calleeClassMethod
     * @throws Exception
     */
    public void addEdge(String callerClassMethod, String calleeClassMethod) throws Exception {
        ClassMethod caller = new ClassMethod(callerClassMethod);
        ClassMethod callee = new ClassMethod(calleeClassMethod);

        if (!graph.containsVertex(caller)) {
            graph.addVertex(caller);
        }
        if (!graph.containsVertex(callee)) {
            graph.addVertex(callee);
        }

        // Adds a directed edge: caller -> callee
        graph.addEdge(caller, callee);
    }

    /**
     * Returns a list of all method invocations made from a given ClassMethod instance
     *
     * @param fromClassMethod Represents the Edge's source
     * @return All edge targets given fromClassMethod as an edge source
     */
    public List<ClassMethod> getMethodCallsFrom(ClassMethod fromClassMethod) {
        List<ClassMethod> methodCalls = new ArrayList<>();

        // If Vertex/Node does not exist, catch the exception and return an empty list
        try {
            for (DefaultEdge edge : graph.outgoingEdgesOf(fromClassMethod)) {
                methodCalls.add(graph.getEdgeTarget(edge));
            }
        } catch (IllegalArgumentException ex) {
            // No methods calls found
            return new ArrayList<>();
        }

        return methodCalls;
    }

    /**
     * Returns a list of all method invocations made to a given ClassMethod instance
     *
     * @param fromClassMethod Represents the Edge's source
     * @return All edge targets given fromClassMethod as an edge source
     */
    public List<ClassMethod> getMethodCallsTo(ClassMethod fromClassMethod) {
        List<ClassMethod> methodCalls = new ArrayList<>();

        // If Vertex/Node does not exist, catch the exception and return an empty list
        try {
            for (DefaultEdge edge : graph.incomingEdgesOf(fromClassMethod)) {
                methodCalls.add(graph.getEdgeTarget(edge));
            }
        } catch (IllegalArgumentException ex) {
            // No methods calls found
            return new ArrayList<>();
        }

        return methodCalls;
    }

    public String toString() {
        StringBuilder output = new StringBuilder();

        for (DefaultEdge edge : graph.edgeSet()) {
            ClassMethod source = graph.getEdgeSource(edge);
            ClassMethod target = graph.getEdgeTarget(edge);
            double weight = graph.getEdgeWeight(edge);
            output.append(source).append(" -> ").append(target).append(" (weight: ").append(weight).append(")\n");
        }

        return output.toString();
    }

    public int size() {
        return graph.vertexSet().size();
    }
}

File: src/main/java/com/mkonst/analysis/MethodCallGraph.kt
package com.mkonst.analysis

import com.mkonst.types.ClassMethod
import org.jgrapht.Graph
import org.jgrapht.graph.DefaultDirectedGraph
import org.jgrapht.graph.DefaultEdge

class MethodCallGraph {
    // Create a directed graph
    val graph: Graph<ClassMethod, DefaultEdge> = DefaultDirectedGraph(DefaultEdge::class.java)

    /**
     * Converts the given Strings into ClassMethod instances and appends a new edge. If the graph does not contain
     * the given methods are Vertices, new vertices will be generated
     *
     */
    @Throws(Exception::class)
    fun addEdge(callerClassMethod: String, calleeClassMethod: String) {
        val caller = ClassMethod(callerClassMethod)
        val callee = ClassMethod(calleeClassMethod)

        if (!graph.containsVertex(caller)) {
            graph.addVertex(caller)
        }
        if (!graph.containsVertex(callee)) {
            graph.addVertex(callee)
        }

        // Adds a directed edge: caller -> callee
        graph.addEdge(caller, callee)
    }

    /**
     * Returns a list of all method invocations made from a given ClassMethod instance
     *
     * @param fromClassMethod Represents the Edge's source
     * @return All edge targets given fromClassMethod as an edge source
     */
    fun getMethodCallsFrom(fromClassMethod: ClassMethod): MutableList<ClassMethod> {
        val methodCalls: MutableList<ClassMethod> = mutableListOf()

        // If Vertex/Node does not exist, catch the exception and return an empty list
        try {
            for (edge in graph.outgoingEdgesOf(fromClassMethod)) {
                methodCalls.add(graph.getEdgeTarget(edge))
            }
        } catch (ex: IllegalArgumentException) {
            // No methods calls found
            return mutableListOf()
        }

        return methodCalls
    }

    override fun toString(): String {
        val output = StringBuilder()

        for (edge in graph.edgeSet()) {
            val source = graph.getEdgeSource(edge)
            val target = graph.getEdgeTarget(edge)
            val weight = graph.getEdgeWeight(edge)
            output.append(source).append(" -> ").append(target).append(" (weight: ").append(weight).append(")\n")
        }

        return output.toString()
    }

    fun size(): Int {
        return graph.vertexSet().size
    }
}
